tempesta_fw/hpack.c

```diff
@@ -3655,6 +2742,253 @@ write_int(unsigned long index, unsigned short max, unsigned short mask,
+			if (unlikely(dst_end - (char *)dst == 0))
+				return -EINVAL;
+			*(u8 *)dst = (u8)last_word;
+		}
+	}
+
+#undef JOIN_BITS
+#undef write_bytes
+
+done:
+	return 0;
+}
+
+static TfwStr *
+tfw_huffman_encode_string(TfwStr *str, TfwPool *pool)
+{
+	unsigned long enc_len = tfw_huffman_encode_string_len(str);
+	TfwStr *encoded;
+	int r;
+
+	if (!enc_len)
+		return ERR_PTR(-EINVAL);
+	encoded = tfw_pool_alloc(pool, sizeof(TfwStr) + enc_len);
+	if (!encoded)
+		return ERR_PTR(-ENOMEM);
+
+	TFW_STR_INIT(encoded);
+	encoded->data = (char *)(encoded + 1);
+	encoded->len = enc_len;
+
+	r = tfw_huffman_encode_copy(str, encoded);
+
+	return r ? ERR_PTR(r) : encoded;
+}
+
+static int
+tfw_hpack_str_add_raw(TfwHttpTransIter *mit, TfwStr *str, bool in_huffman)
+{
+	int r = 0;
+	TfwHPackInt len;
+	TfwStr len_str = { 0 };
+	unsigned short mask = in_huffman ? 0x80 : 0x0;
+
+	write_int(str->len, 0x7F, mask, &len);
+	len_str.data = len.buf;
+	len_str.len = len.sz;
+
+	r = tfw_h2_msg_rewrite_data(mit, &len_str, mit->bnd);
+	if (unlikely(r))
+		return r;
+
+	return tfw_h2_msg_rewrite_data(mit, str, mit->bnd);
+}
+
+static int
+tfw_hpack_str_expand_raw(TfwHttpTransIter *mit, TfwMsgIter *it,
+			 struct sk_buff **skb_head, TfwStr *str,
+			 bool in_huffman)
+{
+	int r;
+	TfwHPackInt len;
+	TfwStr len_str = { 0 };
+	unsigned short mask = in_huffman ? 0x80 : 0x0;
+
+	write_int(str->len, 0x7F, mask, &len);
+	len_str.data = len.buf;
+	len_str.len = len.sz;
+
+	r = tfw_http_msg_expand_data(it, skb_head, &len_str);
+	if (unlikely(r))
+		return r;
+	mit->acc_len += len_str.len;
+
+	r = tfw_http_msg_expand_data(it, skb_head, str);
+	if (unlikely(r))
+		return r;
+	mit->acc_len += str->len;
+
+	return 0;
+}
+
+/*
+ * Family of functions to add new or append of h2 response. Huffman encoding
+ * is never used due to performance considerations:
+ * - We must write first the size of the string, it has variable size, and
+ *   string encoded with Huffman codes may be shorter or longer than original
+ *   string. The size will be known only after the string is completely encoded.
+ * - We can't encode @str in-place in @str: due to variable-sized coding, we
+ *   can't overwrite symbol-by symbol in place. But this is also not possible
+ *   due to @str origin, some strings are predefined strings from const memory
+ *   region, others are shared for all messages to the same vhost.
+ * - For skb management we need to have exact data size before using API.
+ *
+ * Since these reasons we can't encode @str on the fly or encode it while
+ * writing to skb, so allocate a new string and copy @str encoded value there.
+ * It heavily affect performance, since we have two allocations and two copy
+ * operations here. We decided to keep the code, but our core use cases shows
+ * no performance improvement opportunities.
+ *
+ * TODO: We are free to choose encoding when we're adding a new header, but
+ * we have to (?) preserve original encoding when modifying already existent
+ * headers.
```

Comment:
According to RFC we are free to choose encoding (static/dynamic/Huffman) when modifying already existent headers too (e.g. in cases of HTTP/1.1=>HTTP/2  or HTTP/2=>HTTP/2 response proxy). Thus, this is a matter of our implementation decisions. E.g. in future implementation of HTTP/2=>HTTP/2 request forwarding it is better to leave the Huffman-encoding as is from the client request for the better performance. But yes - for the responses forwarding the Huffman-encoding has no any sense for us in context of performance (neither for added headers nor for existing ones).

Reply:
Thank you for clearing up, I will add this note no comment